File: php/46. Permutations.php
<?php

use LDAP\Result;

class Solution
{
    public $result = [];
    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    function permute($nums)
    {
        $this->dfs([], $nums);
        // print_r($this->result);
    }

    function dfs($path, $nums)
    {
        if (count($nums) == 0) {
            $this->result[] = $path;
        }

        for ($i = 0; $i < count($nums) - 1; $i++) {
            $path[] = $nums[$i];
            $v = $nums[$i];
            unset($nums[$i]);
            $this->dfs($path, $nums);
            $nums[$i] = $v;
        }
    }
}

$test = new Solution();
$nums = [1, 2, 3];
$test->permute($nums);

File: python/46. Permutations.py
class Solution:
    def permute(self, nums):
        res = []
        self.dfs(nums, [], res)
        return res

    def dfs(self, nums, path, res):
        if not nums:
            res.append(path)
            # return # backtracking
        for i in range(len(nums)):
            print(res)
            self.dfs(nums[:i] + nums[i + 1 :], path + [nums[i]], res)


nums = [1, 2, 3]
test = Solution()
test.permute(nums)
